### PR TITLE
1199: Enforce maintainer approval in JBS before allowing to integrate backports into updates projects

### PR DESCRIPTION
**Pull request by zhaosongzs**
As Erik said in the description of this issue, currently, this issue only cares about tracking approval labels in the related bugs.

If a repository is set up with the "approval" configuration, pull requests in that repository will require the maintainer's approval in JBS. Otherwise, the pull request will not be considered ready. 

Erik has also provided a design outlining how to configure the "approval" for a repository.

```
The simple case, where the labels are the same for every branch in a repository: 

"approval": { 
  "request": "jdk17u-fix-request", 
  "approved": "jdk17u-fix-yes", 
  "rejected": "jdk17u-fix-no", 
} 

To reduce the need for changing multiple strings when copying a configuration for a new repository, there is an optional "prefix" field: 

"approval": { 
  "prefix": "jdk17u-fix-", 
  "request": "request", 
  "approved": "yes", 
  "rejected": "no", 
} 

When there are multiple branches with different labels, having the prefix set per branch can help reduce the size of the configuration significantly: 

"approval": { 
  "request": "-critical-request", 
  "approved": "-critical-approved", 
  "rejected": "-critical-rejected", 
  "branches": [ 
    "jdk20\\.0\\.1": { "prefix": "CPU23_04" } 
  ] 
} 
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1199](https://bugs.openjdk.org/browse/SKARA-1199): Enforce maintainer approval in JBS before allowing to integrate backports into updates projects (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1537/head:pull/1537` \
`$ git checkout pull/1537`

Update a local copy of the PR: \
`$ git checkout pull/1537` \
`$ git pull https://git.openjdk.org/skara.git pull/1537/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1537`

View PR using the GUI difftool: \
`$ git pr show -t 1537`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1537.diff">https://git.openjdk.org/skara/pull/1537.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1537#issuecomment-1610305331)

### CODE REVIEW

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 1]
```diff
@@ -677,11 +695,25 @@ private String getStatusMessage(PullRequestCheckIssueVisitor visitor,
...
+                                String targetRef = pr.targetRef();
+                                var labels = issueTrackerIssue.get().labelNames();
+                                if (labels.contains(approval.rejectedLabel(targetRef))) {
+                                    status = "Rejected";
+                                } else if (labels.contains(approval.approvedLabel(targetRef))) {
+                                    status = "Approved";
+                                } else if (labels.contains(approval.requestedLabel(targetRef))) {
+                                    status = "Requested";
+                                }
+                                if (!status.isEmpty() && !status.isBlank()) {
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:712]: Checking just for empty should be enough here.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java [hunk 2]
```diff
@@ -179,6 +179,7 @@ String getIssueMetadata(String prBody) {
                             issueData.append(properties.get("priority").asString());
                             issueData.append(properties.get("issuetype").asString());
                         }
+                        issueData.append(String.join("", issue.labelNames()));
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java:182]: This is currently adding another query to the issue tracker, but that can actually be fixed, and I think it's worth doing that in this patch. You can override `Issue::labelNames` in `JiraIssue` and just read the names from the json. There is a labels field with the names in it. I had this in a local patch for a while but it didn't fit with the thing I was working on at the time, and then I forgot.

**Comment by erikj79** [comment]
When a PR is ready with everything but approval, we should also update the status comment with a helpful message linking to the approval process. This is the same comment where the "ready to integrate" message is posted. That link needs to be configurable.

**Comment by erikj79** [comment]
> > When a PR is ready with everything but approval, we should also update the status comment with a helpful message linking to the approval process. This is the same comment where the "ready to integrate" message is posted. That link needs to be configurable.
> 
> Do you mean the `mergeReadyComment`?(I saw your comment in the issue description) But currently, the maintainer approval is a `progress`. The PR would not be considered ready without maintainer approval label in the related issue, and the mergeReadyComment would not be posted unless the PR is considered ready.

That's what I meant. This would mean that we post the comment earlier and that we should rename the concept of that comment to something like `nextStepComment`. But maybe we shouldn't use the same comment, but instead introduce a new `approvalNeededComment`. Otherwise there is risk that it gets lost in the list of comments after a while. This new comment would be added when only the approval is left to do. Later in [SKARA-1949](https://bugs.openjdk.org/browse/SKARA-1949), the contents of this comment would be expanded to contain instructions for how to apply for maintainer approval using Skara commands. For now it can only reference the process documentation.

**Comment by zhaosongzs** [comment]
> > > When a PR is ready with everything but approval, we should also update the status comment with a helpful message linking to the approval process. This is the same comment where the "ready to integrate" message is posted. That link needs to be configurable.
> > 
> > 
> > Do you mean the `mergeReadyComment`?(I saw your comment in the issue description) But currently, the maintainer approval is a `progress`. The PR would not be considered ready without maintainer approval label in the related issue, and the mergeReadyComment would not be posted unless the PR is considered ready.
> 
> That's what I meant. This would mean that we post the comment earlier and that we should rename the concept of that comment to something like `nextStepComment`. But maybe we shouldn't use the same comment, but instead introduce a new `approvalNeededComment`. Otherwise there is risk that it gets lost in the list of comments after a while. This new comment would be added when only the approval is left to do. Later in [SKARA-1949](https://bugs.openjdk.org/browse/SKARA-1949), the contents of this comment would be expanded to contain instructions for how to apply for maintainer approval using Skara commands. For now it can only reference the process documentation.

Ah, got it! Would add the comment.

**Comment by zhaosongzs** [comment]
> > > When a PR is ready with everything but approval, we should also update the status comment with a helpful message linking to the approval process. This is the same comment where the "ready to integrate" message is posted. That link needs to be configurable.
> > 
> > 
> > Do you mean the `mergeReadyComment`?(I saw your comment in the issue description) But currently, the maintainer approval is a `progress`. The PR would not be considered ready without maintainer approval label in the related issue, and the mergeReadyComment would not be posted unless the PR is considered ready.
> 
> That's what I meant. This would mean that we post the comment earlier and that we should rename the concept of that comment to something like `nextStepComment`. But maybe we shouldn't use the same comment, but instead introduce a new `approvalNeededComment`. Otherwise there is risk that it gets lost in the list of comments after a while. This new comment would be added when only the approval is left to do. Later in [SKARA-1949](https://bugs.openjdk.org/browse/SKARA-1949), the contents of this comment would be expanded to contain instructions for how to apply for maintainer approval using Skara commands. For now it can only reference the process documentation.

And now I am thinking why we only post `approvalNeededComment` when a PR is ready with everything but approval? What about posting `approvalNeededComment` when approval is missing?

**Comment by erikj79** [comment]
> And now I am thinking why we only post `approvalNeededComment` when a PR is ready with everything but approval? What about posting `approvalNeededComment` when approval is missing? In this way, user could see the instructions earlier and they would have the time to apply for approval.

I think we should refrain from posting it early because the formal approval process shouldn't start until the PR is reviewed. Posting the comment early might encourage users to post approval requests early, which will annoy the maintainers.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 3]
```diff
@@ -1490,4 +1493,38 @@ private String describe(PullRequest pr) {
...
+        }
+        StringBuilder messageBuilder = new StringBuilder();
+        if (existingUnapproved) {
+            messageBuilder.append("⚠️  @").append(pr.author().username())
+                    .append(" There are still some issues that have not received maintainer approval.")
+                    .append("\n")
+                    .append("Please follow the instruction here to get maintainer approval: ")
+                    .append("[Requesting push approval for fixes](https://openjdk.org/projects/jdk-updates/approval.html)");
+        } else {
+            messageBuilder.append("@").append(pr.author().username()).append(" All the issues have already got the maintainer approval!");
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1517]: When looking at this, I think I'm changing my mind a bit around this. I think we should just post the "approval needed" comment once and not edit it with progress. It's enough to have progress in the pr body.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 4]
```diff
@@ -1497,34 +1508,15 @@ private boolean approvalNeeded() {
...
         if (existing.isPresent()) {
-            if (!existing.get().body().equals(message)) {
-                pr.updateComment(existing.get().id(), message);
-            }
-        } else {
-            pr.addComment(message);
+            return;
         }
+        String message = "⚠️  @" + pr.author().username() +
+                " This change has now been reviewed and requires maintainer [approval](" + approval.documentLink() + ")." +
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1517]: I did some more thinking about how this should be formulated. I don't think we should mention the review state here either. 
```suggestion
                " This change is now ready for maintainer [approval](" + approval.documentLink() + ")." +
```

File: bots/pr/src/test/java/org/openjdk/skara/bots/pr/ApprovalTests.java [hunk 5]
```diff
@@ -31,19 +31,20 @@
 public class ApprovalTests {
     @Test
     void simple() {
-        Approval approval = new Approval("", "jdk17u-fix-request", "jdk17u-fix-yes", "jdk17u-fix-no");
+        Approval approval = new Approval("", "jdk17u-fix-request", "jdk17u-fix-yes", "jdk17u-fix-no", "https://test.openjdk.org");
```
Comment (erikj79) [inline bots/pr/src/test/java/org/openjdk/skara/bots/pr/ApprovalTests.java:34]: I recommend using something with `example.com` for test URL.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 4 above)
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1517]:
```suggestion
                " This change is now ready for you to apply for maintainer [approval](" + approval.documentLink() + ")." +
```

File: bots/pr/src/test/java/org/openjdk/skara/bots/pr/ApprovalTests.java (hunk 5 above)
Comment (zhaosongzs) [reply bots/pr/src/test/java/org/openjdk/skara/bots/pr/ApprovalTests.java:34]: Sure, will fix it.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 4 above)
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java]: Did you see my last suggestion here? I think we need to be clear that the ball is in the hands of the PR owner.

Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java]: Sorry, I missed it. Will fix it soon.

**Review by erikj79 — APPROVED** [review]

**Comment by zhaosongzs** [comment]
Thanks for the review!
/integrate

**Pull request closed without merging**